### PR TITLE
Explicitly initialize prevector _union

### PR DESCRIPTION
**Pull request by Empact**

### CODE REVIEW

File: src/prevector.h [hunk 1]
```diff
@@ -163,6 +163,22 @@ class prevector {
     const T* indirect_ptr(difference_type pos) const { return reinterpret_cast<const T*>(_union.indirect) + pos; }
     bool is_direct() const { return _size <= N; }
 
+    /**
+     * @pre _size must be set to the desired size
+     */
+    union direct_or_indirect init_union(void) {
+        if (is_direct()) {
+            return {};
+        } else {
```
Comment (Empact) [inline src/prevector.h]: Compare with:
https://github.com/bitcoin/bitcoin/blob/271b379e636afa419c5208cb462c07090490266c/src/prevector.h#L185-L192

File: src/prevector.h [hunk 2]
```diff
@@ -163,6 +163,22 @@ class prevector {
...
+            union direct_or_indirect indirect;
+            indirect.indirect = static_cast<char*>(malloc(((size_t)sizeof(T)) * _size));
+            assert(indirect.indirect);
+            indirect.capacity = _size;
+            _size += N + 1;
+            return indirect;
+        }
```
Comment (Empact) [inline src/prevector.h]: Compare with:
https://github.com/bitcoin/bitcoin/blob/271b379e636afa419c5208cb462c07090490266c/src/prevector.h#L185-L192

File: src/prevector.h [hunk 3]
```diff
@@ -248,32 +264,24 @@ class prevector {
 
     prevector() : _size(0), _union{{}} {}
 
-    explicit prevector(size_type n) : _size(0) {
-        resize(n);
+    explicit prevector(size_type n) : _size(n), _union{init_union()} {
```
Comment (sipa) [inline src/prevector.h]: This is wrong. The `_size` field does not directly encode the size (see the comments above the class).

Comment (Empact) [reply src/prevector.h]: init_union adds the `N + 1` when appropriate.

Comment (sipa) [reply src/prevector.h]: Yuck.

Comment (Empact) [reply src/prevector.h]: Fair enough, moved it out of the initializer.

**Comment by practicalswift** [comment]
Concept ACK

This fixes the compiler warning described in #13902

**Comment by Empact** [comment]
Pared the PR down to just explicitly initializing the _struct

**Comment by practicalswift** [comment]
The updated version still fixes the compiler warning.

ACK f955a50384ad53e6d1e89f373905ea8dabc46e93

**Comment by Empact** [comment]
Rebased for  #14030

**Comment by practicalswift** [comment]
utACK 1d9aa008d6e043c29c3c5b030a6d04278aea233b

**Comment by ken2812221** [comment]
utACK 1d9aa008d6e043c29c3c5b030a6d04278aea233b

**Pull request merged**